File: lib/src/main/java/com/sd/lib/cache/simple/SimpleMultiObjectCache.java
package com.sd.lib.cache.simple;

import android.text.TextUtils;

import com.sd.lib.cache.Cache;
import com.sd.lib.cache.CacheInfo;
import com.sd.lib.cache.handler.ObjectHandler;

public class SimpleMultiObjectCache<T> implements Cache.MultiObjectCache<T>
{
    private final ObjectHandler mObjectHandler;
    public final Class<T> mObjectClass;

    public SimpleMultiObjectCache(CacheInfo info, Class<T> objectClass)
    {
        if (objectClass == null)
            throw new NullPointerException("objectClass is null");

        mObjectClass = objectClass;
        mObjectHandler = new ObjectHandler(info)
        {
            @Override
            protected String getKeyPrefix()
            {
                return "multi_object_";
            }
        };
    }

    @Override
    public boolean put(String key, T value)
    {
        if (TextUtils.isEmpty(key))
            throw new IllegalArgumentException("key is null or empty");

        key += mObjectClass.getName();
        return mObjectHandler.putCache(key, value);
    }

    @Override
    public T get(String key)
    {
        if (TextUtils.isEmpty(key))
            throw new IllegalArgumentException("key is null or empty");

        key += mObjectClass.getName();
        return (T) mObjectHandler.getCache(key, mObjectClass);
    }

    @Override
    public boolean remove(String key)
    {
        if (TextUtils.isEmpty(key))
            throw new IllegalArgumentException("key is null or empty");

        key += mObjectClass.getName();
        return mObjectHandler.removeCache(key);
    }

    @Override
    public boolean contains(String key)
    {
        if (TextUtils.isEmpty(key))
            throw new IllegalArgumentException("key is null or empty");

        key += mObjectClass.getName();
        return mObjectHandler.containsCache(key);
    }
}

File: lib/src/main/java/com/sd/lib/cache/simple/SimpleMultiObjectCache.kt
package com.sd.lib.cache.simple

import com.sd.lib.cache.Cache.MultiObjectCache
import com.sd.lib.cache.CacheInfo
import com.sd.lib.cache.handler.impl.ObjectHandler

internal class SimpleMultiObjectCache<T>(
    cacheInfo: CacheInfo,
    val objectClass: Class<T>,
) : MultiObjectCache<T> {
    private val _objectHandler = ObjectHandler(cacheInfo, "multi_object")

    private fun transformKey(key: String): String {
        require(key.isNotEmpty()) { "key is empty" }
        return "${objectClass.name}_$key"
    }

    override fun put(key: String, value: T?): Boolean {
        if (key.isEmpty()) return false
        if (value == null) return false
        return _objectHandler.putCache(transformKey(key), value, objectClass)
    }

    @Suppress("UNCHECKED_CAST")
    override fun get(key: String): T? {
        if (key.isEmpty()) return null
        val cache = _objectHandler.getCache(transformKey(key), objectClass) ?: return null
        return cache as T
    }

    override fun remove(key: String): Boolean {
        if (key.isEmpty()) return false
        return _objectHandler.removeCache(transformKey(key))
    }

    override fun contains(key: String): Boolean {
        if (key.isEmpty()) return false
        return _objectHandler.containsCache(transformKey(key))
    }
}